File: python/src/xmlai/llm.py
from typing import List, Union
from . import generate_xml_prompt


def anthropic_prompt(prompt: dict, response_root_tag="response") -> dict:
    """
    This function generates a prompt for Anthropic's Claude API.

    Parameters:
        prompt (dict): The prompt to be generated. This should be a dictionary representation of the XML string.
        response_root_tag (str): The root tag of the response XML. Defaults to "response".
    Returns:
        dict: A dictionary containing the prompt and stop sequences, ready to be passed to the anthropic API.

    Example:
        >>> completion = anthropic.completions.create(
        ...     model="claude-2",
        ...     max_tokens_to_sample=300,
        ...     **claude_prompt(
        ...         {
        ...             "question": "what is the answer to the ultimate question of life?",
        ...             "reference": "The Hitchhiker's Guide to the Galaxy",
        ...         },
        ...         "answer",
        ...     ),
        ... )
        >>> completion.completion
        '42'
    """

    return {
        "prompt": f"\n\nHuman:{generate_xml_prompt(prompt)}\n\nAssistant:<{response_root_tag}>",
        "stop_sequences": [f"</{response_root_tag}>"],
    }


def openai_chat_prompt(
    messages: List[dict],
    response_root_tag="response",
    custom_system_prefix: Union[None, str] = None,
) -> dict:
    """
    This function generates a prompt for OpenAI's API.

    Parameters:
        messages (List[dict]): A list of messages in OpenAI format. Content is optional for system prompt.
        response_root_tag (str): The root tag of the response XML. Defaults to "response".

    Returns:
        dict: A dictionary containing the prompt and stop sequences, ready to be passed to the OpenAI API.
    """

    if custom_system_prefix is None:
        custom_system_prefix = f"Respond in XML. Your must start your responses with <{response_root_tag}>\n\n"

    for message in messages:
        if "content" in message and isinstance(message["content"], dict):
            message["content"] = generate_xml_prompt(message["content"])
        if message["role"] == "system":
            if "content" not in message:
                message["content"] = ""
            message["content"] = custom_system_prefix + message["content"]

    return {
        "messages": messages,
        "stop": [f"</{response_root_tag}>"],
    }

File: ts/src/llm.ts
import { generateXmlPrompt } from ".";

/** JSON object with only strings as keys/values. */
export type StringJson = string | { [key: string]: StringJson };

type OpenAIMessage = {
  role: "system" | "user" | "assistant" | "function";
  content?: string;
  name?: string;
};

type OpenAIMessageJson = Omit<OpenAIMessage, "content"> & {
  content?: StringJson;
};

interface OpenAIPrompt {
  messages: OpenAIMessageJson[];
  response_root_tag?: string;
  custom_system_prefix?: string;
}

/**
 * Generates an Anthropic prompt for the given XML prompt and response root tag.
 *
 * @param prompt The XML prompt.
 * @param response_root_tag The root tag for the response.
 * @returns a JSON prompt and stop sequences that can be passed into Anthropic
 */
export function anthropicPrompt(
  prompt: StringJson,
  response_root_tag = "response"
) {
  return {
    prompt: `\n\nHuman:${generateXmlPrompt(
      prompt
    )}\n\nAssistant:<${response_root_tag}>`,
    stop_sequences: [`</${response_root_tag}>`],
  };
}

/**
 * Generate an OpenAI Chat Prompt from (optionally) JSON messages
 *
 * @param messages The array of OpenAIMessageJson objects. The content of these messages may be JSON.
 * @param response_root_tag The root tag for the response.
 * @param custom_system_prefix The custom system prefix for the response.
 * @returns The JSON prompt and stop sequences that can be passed into OpenAI.
 */
export function openaiChatPrompt({
  messages,
  response_root_tag = "response",
  custom_system_prefix,
}: OpenAIPrompt) {
  if (custom_system_prefix === undefined) {
    custom_system_prefix = `Respond in XML. Your must start your responses with <${response_root_tag}>\n\n`;
  }

  for (const message of messages) {
    if ("content" in message && typeof message["content"] === "object") {
      message["content"] = generateXmlPrompt(message["content"]);
    }
    if (message["role"] === "system") {
      if (!("content" in message)) {
        message["content"] = "";
      }
      message["content"] = custom_system_prefix + message["content"];
    }
  }

  return {
    messages: messages as OpenAIMessage[],
    stop: [`</${response_root_tag}>`],
  };
}
